pkg/cmd/util/tokencmd/challenging_client.go

```diff
@@ -39,19 +39,26 @@ func (client *challengingClient) Do(req *http.Request) (*http.Response, error) {
 
 			uDefaulted := len(username) > 0
 			pDefaulted := len(password) > 0
+			shouldRetry := false
 
-			if !(uDefaulted && pDefaulted) {
+			if uDefaulted && pDefaulted {
+				shouldRetry = true
+			} else if client.reader != nil {
 				fmt.Printf("Authenticate for \"%v\"\n", realm)
 				if !uDefaulted {
 					username = util.PromptForString(client.reader, "Username: ")
+					shouldRetry = true
 				}
 				if !pDefaulted {
 					password = util.PromptForPasswordString(client.reader, "Password: ")
+					shouldRetry = true
 				}
 			}
 
-			client.delegate.Transport = kclient.NewBasicAuthRoundTripper(username, password, client.delegate.Transport)
-			return client.Do(resp.Request)
+			if shouldRetry {
```

Comment:
Since it no longer infinitely loops, can't you just always try with what ever username/password you have?  I don't see why the bool is needed.